README

GERB2TIFF 
====================================================================
utility to generate raster graphics bitmap from Gerber RS274X files.


Package contents
--------------------------

File/path                Description
----------------------------------------------------------
aclocal.m4            
ChangeLog            List of changes to each version release
configure
configure.ac
COPYING              GNU General Publuc Lisence
INSTALL              General Installation Instructions
Makefile.am         
Makefile.in
NEWS
README   		     This file 
build-aux/           Auxilary programs for configure
src/                 Source files
html/                Documentation
examples/            Examples of RS274X gerber files 




Documention
--------------------------

All the documentation is in HTML in the html directory.  
To view the documentation point your favourite WWW viewer at html/index.html;


Install
--------------------------
See INSTALL for general install instructions.
Basically what you want to do is:

    % ./configure
    % make
    % su
    % make install


Dependencies
--------------------------
Libraries:

    libtiff - version 3.9.1 or later. Available from http://www.libtiff.org/
    
    

Use and Copyright
-------------------------

Copyright (c) 2009 Adam Seychell
 
This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation. This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
details. A copy of the GNU General Public License is included in the file
COPYING of this package.
 



ChangeLog
Proposed changes to next revision:
------------------------------------------------------------------------------
 * correctly implement blank holes in standard apertures flashes
 



Description of revision changes : (release date)

Version  1.2      (6-Feb-2011)
------------------------------------------------------------------------------
* Added IR (Image Rotation) parameter support.
* Added IJ (Image Justify) parameter support.
* Changed build process for GNU autotool built system.
  See INSTALL for instruction on compiling.
  The Windows 32bit executalbe is in separated archive, 
  gerb2tiff-{version}-win32-bin.zip
* removed redundant sub-minor version numbers    



Version  1.0.9      (20-Dec-2010)
------------------------------------------------------------------------------
* bug fix: Polygon drawing code completely rewritten. This fixed problems with 
  arcs and line mismatching. Also fixed the problem of occurring rouge horizontal 
  scan lines. 
* bug fix: Aperture primitive 21 (Line Centre) had centre position located at 
  upper left instead of lower left.
* Improved bitmap rendering speed.

Version  1.0.8      (8-Oct-2009)
------------------------------------------------------------------------------
* bug fix: fixed crash when multiple input gerber files specified.

Version  1.0.7      (3-Oct-2009)
------------------------------------------------------------------------------
* bug fixed: arcs with equal start and ending points were not drawn.

Version  1.0.6      (16-July-2009)
------------------------------------------------------------------------------
* bug fixed: The width of traces were incorrectly scaled to the square of the
  scale factor.
* bug fixed: IPNEG and IPPOS parameters were not being recognised.
* Implemented MI (Mirror Axis) parameter
* Implemented IF (Include File) parameter. note: nested includes are allowed, and 
  recursive includes are not safeguarded. Error reporting does not contain include 
  file information. These two problems can be fixed in future versions. 
* Arcs are now supported in polygon drawing
* Improved robustness in reading coordinate data. Can handle excess digits than 
  designated in FS parameter.
* Improved arc drawing algorithm yet again. Added correction mechanism for 
  mismatched arc centres in 360 degree circular mode. In single quadrant, arc 
  centres are always relative to the tool location prior to G03/G02 command, by an 
  offset given by I,J. In single quadrant mode arc radius = sqrt(I*I + J*J).
  Arc with zero radius are replaced with line segments. 
  (Gerber arcs are horrible to implement)  

Version  1.0.5      (10-July-2009)
------------------------------------------------------------------------------
* bug fixed: The last vertex in G36 command polygons was ignored. This caused 
  polygons that did not have the last vertex matching the first vertex to be drawn 
  incorrectly.
* bug fixed: An aperture change followed by any drawing within the same 
  command block would not draw the first line/arc correctly.
* bug fixed: Any command data block caused drawing to occur if the lamp state 
  was on. For example, the command block 'G37*' would draw to current X Y position
  if lamp was on. The correct method draws only if a command block 
  contains any one of the following commands: X, Y, I, J or D1 (lamp on) and is 
  not in polygon mode.     
* improved error reporting on macro definitions              


Version  1.0.4      (9-July-2009)
------------------------------------------------------------------------------
* bug fixed: Consecutive ADD data blocks were not recognised inside a single '%' block.              

Version  1.0.3      (4-July-2009)
------------------------------------------------------------------------------
* added options, --grow-mm, --boarder-pixels, --strip-rows
* verbose mode reports bitmap origin location.
* multiple Gerber files on command line are overlayed to a single bitmap.
  


Version  1.0.2      (22-June-2009)
------------------------------------------------------------------------------
* Fixed several bugs with arc drawing
* Added support linear line drawing using the standard Rectangle aperture.
* Added options for image scaling.
* Improved error reporting
* Improved command line interface.
* Capable of reading Gerber data form standard input

Version  1.0.0beta  (5-June-2009)
------------------------------------------------------------------------------
* The project was renamed to gerb2tiff
* A complete rewrite of the source code. 
   The image is now rendered entirely by polygons
   Uses bison and flex for the parser
* Close to full implementation of the full RS-274X standard
* Improved speed and memory usage.  
 
Version  0.74  (19-October-2007)

src/gerber.h
/*
	copyright (c), 2001 Adam Seychell.


This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/
#ifndef GERBER_H_
#define GERBER_H_


#include <iostream>
#include <sstream>
#include <string>
#include <complex>
#include <vector>
#include <list>
#include <math.h>
#include <ctype.h>
#include <getopt.h>
#include <tiffio.h>
#include <string.h>

#include "polygon.h"
#include "apertures.h"


using namespace::std;

//#define DEBBUG

#ifdef DEBBUG
#define DEBUGPRINT(...) printf(__VA_ARGS__);
#else
#define DEBUGPRINT(...)
#endif




typedef struct T_Arc
{
	Point centre;
	Point stopped;
	double radius;
	double start;
	double end;
	bool isTooSmall;	// false if arc is too small to plot
} Arc;


extern char *yytext;
extern void yyrestart( FILE *new_file );


class Gerber {


	friend  int yylex(Gerber *gerb);
	friend  int yyparse(Gerber *gerb);
	private:
		enum APETURE_DRAWING_MODE {CIRCLE_CLOCKWISE, CIRCLE_ANTICLOCKWISE, LINEAR_10X, LINEAR_1X, LINEAR_01X, LINEAR_001X, CIRCULAR360, _INVALID_};
		typedef enum {MILLIMETER, INCH, UNDEFINED} Units_t ;

		ostringstream oss;
		const double optScaleY;
		const double optScaleX;
		const double growSize;
   		const double dotsPerInch;
		double coordPrevious[2];
		double scaleFactor[2];
		double imageOffsetPixels[2];		// offset of the image in pixels specified by %IO parameter
		bool layerPolarityClear;
		int coordsDecimals[2];
		int coordsInts[2];
		bool isAxisSwapped;
		bool isOmitLeadingZeroes;
        bool isCoordsAbsolute;
        Units_t units;
		bool isMirrorAaxis;				// negative all A axis coordinate data
		bool isMirrorBaxis;				// negative all B axis coordinate data
		int currentLine;
		enum APETURE_DRAWING_MODE 	drawingMode;
		bool isCircular360;
		bool isPolygonFill;
		bool isLampOn;
		std::list<Aperture>::iterator lastDrawnApertureSelect;
		double lastDrawnX;
		double lastDrawnY;
		double imageRotate;
		bool isDrawingEnabled;

		int warningCount;
		bool isWarnNoApertureSelect;

		std::list<Aperture>::iterator apertureSelect;

		std::vector< NodeT * > 	temporaryParameters;
		string 					temporaryNameMacro;

		std::vector<double>  variables;

		double X;
		double Y;
		double oldX;
		double oldY;
		double I;
		double J;
		double dotsPerUnit();
		const char * unitText()	{	switch (units) { case MILLIMETER: return "mm"; case INCH: return "\""; }	}
		double getCoordinate( char * const text, int axisNumber, bool alwaysAbsolute = false);
		void calculateArc(Arc &arc);
		void process_AD_block(int DCode);
		void process_D_command(int code);
		void process_G_command(int code);
		void processDataBlock();
		void flashAperture(double x, double y);
		void loadDefaults();
        std::list< Aperture > macro_apertures;
        std::list< Aperture > ad_apertures;

        struct StepRpeatBlock
        {
        	string buffer;
        	int X;
        	int Y;
        	double I;
        	double J;
        } repeat;


	public: //---------------------------
		void warning(const char * format, ...);
		bool imagePolarityDark;
		string layerName;
		string imageName;
		string imageFilm;
		std::vector< std::string > messages;
		std::ostringstream errorMessage;			// string of occurring error
		bool isError;

		list<Polygon> polygons;		// Contains a complete polygons list to build an image of this gerber file.
		list<VertexData *> vertexdata;	// Vertices information used by each new polygon requiring a new set of vertices.

		Gerber(FILE * fp_gerb, double ImageDPI, double GrowSize, double optScaleX, double optScaleY);
};


extern void yyerror( Gerber *g, const char * msg);











#endif // GERBER_H_


src/main.cpp
/*
File:	 	main.cpp
Author: 	Adam Seychell
Description:  main function for gerb2tiff program.


	copyright (c), 2001 Adam Seychell.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/

#include <time.h>
#include <iostream>
#include <sstream>
#include <algorithm>
#include <vector>
#include <list>
#include <map>
#include <string>
#include <math.h>
#include <limits.h>
#include <ctype.h>
#include <getopt.h>
#include <tiffio.h>
#include <stdarg.h>
#include <string.h>
#include <config.h>

using namespace std;

#include "polygon.h"
#include "apertures.h"
#include "gerber.h"


unsigned char * DEGUB_bitmap_ptr_end;

unsigned char nbitsTable[256];


const char *help_message=
"gerb2tiff  version "PACKAGE_VERSION" Copyright (c) 2010 by Adam Seychell\n"
"Gerber RS-274X file to raster graphics converter \n"
"\n"
"Usage: gerb2tiff [OPTIONS] [file1] [file2]...\n"
"\n"
"Output control: \n"
"  -a, --area           Show total dark area of TIFF in square centimeters.\n"
"  -q, --quiet          Suppress warnings and non critical messages.\n"
"  -t                   Test only. Process Gerber file without writing TIFF.\n"
"  -o, --output=FILE    Set name of output TIFF to FILE. If gerber-file is\n"
"                       specified then default is <file1>.tiff\n"
"                       This option is required when no gerber-file specified.\n"
"  -v                   Verbose mode, display information while processing\n"
"                       multiple -v increases verbosity. Disables --quiet\n"
"  --help               This help screen\n"
"\n"
"Image options: \n"
"  --boarder-pixels=X   Add a boarder of X pixels around image. Default 0\n"
"  -b, --boarder-mm=X   same as --boarder except X is in millimeters\n"
"  -p, --dpi=X          Number of dots per inch X. Default 2400\n"
"  -n, --negative       Negate image polarity\n"
"  --grow-pixels=X      Expand perimeter of all aperture features by X pixels.\n"
"                       Negative values shrink. Fractional pixels allowed.\n"
"  --grow-mm=X          Same as --grow-pixels except X is in unit millimeters.\n"
"  --strip-rows=N       Specify N rows per strip in TIFF. Default 512\n"
"  --scale-y=FACTOR     Scale image in Y axis by FACTOR. Default 1\n"
"  --scale-x=FACTOR     Scale image in X axis by FACTOR. Default 1\n"
"\n"
"Where file1 file2... are gerber files rendered as overlays to a single bitmap.\n"
"Standard input is read if no gerber files specified and --output is specified.\n"
"Output bitmap is compressed monochrome TIFF.\n"
"\n"
"For latest releases and report bugs visit gerb2tiff home page at:\n"
" http://members.optusnet.com.au/eseychell\n"

;


void show_interval(const char * msg="")
{
	static clock_t  start_clock = clock();
    double cpu_time_used = ((double) (clock() - start_clock)) / CLOCKS_PER_SEC;
	printf("time: %.3f s (%s)\n",cpu_time_used,  msg);
	start_clock = clock();
}

void error(const string &message)
{
	cerr << "gerb2tiff: error: "<<message<<endl;
    exit(1);
}


//***************************************************
// Global variables of plotting parameters
//**************************************************
    double imageDPI = 2400;
    double optRotation;
    bool   optGrowUnitsMillimeters = false;
    bool   optBoarderUnitsMillimeters = false;
    double optBoarder = 0;
    bool  optInvertPolarity = false;
    bool  optTestOnly = false;
    int  optVerbose = 0;
    unsigned rowsPerStrip = 512;
    bool  optShowArea = false;
    bool  optQuiet = false;
    double total_area_cmsq = 0;
	double optGrowSize = 0;
	double optScaleX = 1;
	double optScaleY = 1;
	unsigned int bytesPerScanline;
	unsigned int bitmapBytes;
	unsigned char * bitmap;

//***********************************************************




//**********************************************************
// Optimised horizontal line drawing from x1,y to x2,y in the monochrome bitmap
// polarity specifies how pixels are changed.
// DRAW_ON = line is drawn bits set
// DRAW_OFF = line is drawn bits cleared
// DRAW_REVERSE  = line is drawn bits inverted
//
// global dependencies:	bytesPerScanline, bitmap
//**********************************************************
void horizontalLine( int x1, int x2, unsigned char *buffer, Polarity_t polarity)
{
	if (x1 > x2)
		swap(x1, x2);

	static char fillSingle[64] = {
			0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
			0xC0, 0x40, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
			0xE0, 0x60, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00,
			0xF0, 0x70, 0x30, 0x10, 0x00, 0x00, 0x00, 0x00,
			0xF8, 0x78, 0x38, 0x18, 0x08, 0x00, 0x00, 0x00,
			0xFC, 0x7C, 0x3C, 0x1C, 0x0C, 0x04, 0x00, 0x00,
			0xFE, 0x7E, 0x3E, 0x1E, 0x0E, 0x06, 0x02, 0x00,
			0xFF, 0x7F, 0x3F, 0x1F, 0x0F, 0x07, 0x03, 0x01 };

	static char fillLast[8]  = {0x80, 0xC0, 0xE0, 0xF0, 0xF8, 0xFC, 0xFE, 0xFF};
	static char fillFirst[8] = {0xFF, 0x7F, 0x3F, 0x1F, 0x0F, 0x07, 0x03, 0x01};


	const char b1 = (x1 & 7);
	const char b2 = (x2 & 7);

	unsigned char *px1 = buffer + (x1 >> 3);
	unsigned char *px2 = buffer + (x2 >> 3);

    // left pixel = MSB
    // right pixel = LSB
	switch (polarity)
	{
	case DARK:	// plot line with set bits
		// fill in the pixels at the byte x1, and x2 occupy.
		if (px1 == px2)
		{ // x1 and x2 occupy the same  byte
			*px1 |= fillSingle[ b1 + (b2<<3) ];
		}
		else
		{ // x1 and x2 occupy different bytes
			*px1 |= fillFirst[ b1 ];
			*px2 |= fillLast[ b2 ];
			// fill only the whole bytes in buffer between x1 and x2
			px1++;
			memset(px1, 0xFF, (px2 - px1));
		}
		break;

	case CLEAR:		// plot line with cleared bits

		if (px1 == px2)	// fill in the pixels at the byte x1, and x2 occupy.
		{ // x1 and x2 occupy the same  byte
			*px1 &= ~fillSingle[ b1 + (b2<<3) ];
		}
		else
		{ // x1 and x2 occupy different bytes
			*px1 &= ~fillFirst[ b1 ];
			*px2 &= ~fillLast[ b2 ];
			// fill only the whole bytes in buffer between x1 and x2
			px1++;
			memset(px1, 0x0, (px2 - px1));
		}
		break;

	case XOR: // invert the pixels
		// fill in the pixels at the byte x1, and x2 occupy.
		if (px1 == px2)
		{ // x1 and x2 occupy the same  byte
			*px1 ^= fillSingle[ b1 + (b2<<3) ];
		}
		else
		{ // x1 and x2 occupy different bytes
			*px1 ^= fillFirst[ b1 ];
			*px2 ^= fillLast[ b2 ];
			// XOR only the whole bytes in buffer between x1 and x2 (exclusive)
			px1++;
			while (px1 < px2)
			{
				*px1 ^= 0xFF;
				px1++;
			}
		}
		break;
	}

} // end HorizontalLine()




//---------------------------------------------------------------------------------
int main (int argc, char **argv)
{
	clock_t  start_clock = clock();;
	string inputfile;
	string outputFilename;




	// create 256 element look up table for fast retrieve of
	// number of bits set in numbers 0 to 255.
	// used for calculating positive pixels in the drawn bitmap

	for (int i=0; i < 256; i++)
	{
		nbitsTable[i]=0;
		if ((i&0x01)) nbitsTable[i]++;
		if ((i&0x02)) nbitsTable[i]++;
		if ((i&0x04)) nbitsTable[i]++;
		if ((i&0x08)) nbitsTable[i]++;
		if ((i&0x10)) nbitsTable[i]++;
		if ((i&0x20)) nbitsTable[i]++;
		if ((i&0x40)) nbitsTable[i]++;
		if ((i&0x80)) nbitsTable[i]++;
	}

	//
    // parse the command line
    //
	while (1)
	{
        static struct option long_options[] =
          {
            // These options don't set a flag.
            // We distinguish them by their indices.
            {"dpi",     required_argument, 0, 'p'},
            {"output",  required_argument, 0, 'o'},
            {"negative",no_argument, 	   0, 'n'},
            {"area",    no_argument, 	   0, 'a'},
            {"test",    no_argument, 	   0,  't'},
            {"quiet",   no_argument, 	   0,  'q'},
            {"verbose", no_argument,   	   0,  'v'},
            {"scale-x", required_argument, 0, 1},
            {"scale-y", required_argument, 0, 2},
            {"help",    no_argument, 	   0,  3},
            {"grow-pixels",	required_argument, 0, 4},
            {"grow-mm",	required_argument, 0, 5},
            {"strip-rows",	required_argument, 0, 6},
            {"boarder-mm", required_argument, 0, 'b'},
            {"boarder-pixels", required_argument, 0, 7},
            {"rotation", required_argument, 0, 8},
            {0, 0, 0, 0}
          };
        // getopt_long stores the option index here.
        int option_index = 0;

        int c = getopt_long (argc, argv, "G:b:o:p:atvnq",
							long_options, &option_index);

		if (c == EOF)		break;

		switch (c)
		{

		case 8:
			optRotation = atof(optarg);
		  break;
		case 7:
			optBoarder = atof(optarg);
			optBoarderUnitsMillimeters = false;
		  break;
		case 6:
			rowsPerStrip = atoi(optarg);
		  break;
		case 5:
		  optGrowSize = atof(optarg);
		  optGrowUnitsMillimeters = true;
		  break;
		case 4:
		  optGrowSize = atof(optarg);
		  optGrowUnitsMillimeters = false;
		  break;
		case 3:
			fprintf( stdout,"%s", help_message);
			exit(0);
		case 2:
		  optScaleY = atof(optarg);
		  break;
		case 1:
		  optScaleX = atof(optarg);
		  break;
		case 'p':
		  imageDPI = atof(optarg);
		  break;
		case 'b':
		  optBoarder = atof(optarg);
  		  optBoarderUnitsMillimeters = true;
		  break;
		case 'n':
		  optInvertPolarity = true;
		  break;
		case 't':
		  optTestOnly = true;
		  break;
		case 'a':
			optShowArea = true;
		  break;
		case 'v':
		  optVerbose++;
		  break;
		case 'q':
		  optQuiet = true;
		  break;
		case 'o':
			outputFilename = optarg;
		  break;
		case '?':
		case ':':
		  fprintf (stderr, "Try 'gerb2tiff --help' for more information.\n");
		  return 1;
		  break;
		}
	}


	if (optVerbose > 0 ) optQuiet = false;			// if user wants verbose, then cancel quiet option

	if (imageDPI < 1)		error(string("DPI setting must be >= 1"));
	if (optBoarder < 0)		error(string("boarder setting must be >= 0"));

	// correct the units for some options
	if ( optGrowUnitsMillimeters )
		optGrowSize *= imageDPI/25.4;
	if ( optBoarderUnitsMillimeters )
		optBoarder *= imageDPI/25.4;

    list<Gerber *> gerbers;			// pointer to the list of Gerber object

    bool isStandardInput = false;
	if (optind == argc)
		isStandardInput = true;


	int first_optind = optind;

	for(; optind < argc || isStandardInput; optind++)
	{
		FILE *file = 0;
		if ( isStandardInput )
		{
			file = stdin;
			if (!optTestOnly && outputFilename.empty())
			{
				cerr << "no output or input file specified.\n"
					    "Try 'gerb2tiff --help' for more information.\n";
				return 1;
			}
		}
		else
		{
			inputfile = argv[optind];
			if ( outputFilename.empty())
					outputFilename = inputfile + ".tiff";
			file = fopen( argv[optind], "rb");
			if (file == NULL)
					error( string("cannot open input file ")+inputfile );
			if (!optQuiet)
			{
				if (optind == first_optind)			cout << "gerb2tiff: ";
				if (!optQuiet && optind > first_optind)	cout << "+ ";
				cout << inputfile << " "<<flush;
			}
		}

		gerbers.push_back( new Gerber(file, imageDPI, optGrowSize, optScaleX, optScaleY) );

		if (! isStandardInput)
			fclose(file);

		// print all warning messages
		for (int i=0; i < gerbers.back()->messages.size() && !optQuiet; i++)
		{
			if (i==0) std::cout <<"\n";
			std::cout <<"("<<inputfile<<") "<<gerbers.back()->messages[i]<<endl;	    		// print messages if any
		}
		// print error messages and abort
		if (gerbers.back()->isError)
		{
			std::cout <<"\n("<<inputfile<<") "<<gerbers.back()->errorMessage.str() << endl;
			return 1;		// exit program
		}
		if ( isStandardInput )
			break;

	}


	if (!optTestOnly  && !optQuiet)		cout << "-> "<<outputFilename;
	if (!optQuiet)						cout << endl;


	int miny =  INT_MAX;			// holds min and max dimentions of the occupied gerber images (superimposed)
	int minx =  INT_MAX;
	int maxy =  INT_MIN;
	int maxx =  INT_MIN;
    list<Polygon> globalPolygons;	// Contains polygons created by the all gerbers.

	// group all the polygons
    for (list<Gerber*>::iterator it = gerbers.begin(); it != gerbers.end();  it++)
    {
       globalPolygons.merge((*it)->polygons );
	}


//    for (int i=0; i < 100; i++)
//    {
//    	double radius = 50.0 * rand() / double(RAND_MAX);
//    	double x0  = 2000.0 * rand() / double(RAND_MAX);
//    	double y0  = 2000.0 * rand() / double(RAND_MAX);
//        globalPolygons.push_back(Polygon() );
//        globalPolygons.back().number = 1;
//        globalPolygons.back().vdata->add( 0,0);
//        globalPolygons.back().vdata->add(100,0);
//        globalPolygons.back().vdata->add(100,50);
//        globalPolygons.back().vdata->add( 0,50);
//        globalPolygons.back().vdata->rotate(222 *M_PI/180);
//        globalPolygons.back().vdata->initialise();
//        globalPolygons.back().initialise();
//    }

//    globalPolygons.sort();
//        globalPolygons.push_back(Polygon() );
//        globalPolygons.back().addArc(0, 2*M_PI, 10, 0, 0, false );
//        globalPolygons.back().number = 0;
//        globalPolygons.back().initialise();



	if (globalPolygons.size()==0 )// If nothing to draw then abort with error
		error("no image");


	// find extreme (x,y) coordinates for all polygons
	for (list<Polygon>::iterator it = globalPolygons.begin(); it != globalPolygons.end();  it++)
	{
		if (minx > it->pixelMinX) 	minx = it->pixelMinX;
		if (maxx < it->pixelMaxX) 	maxx = it->pixelMaxX;
		if (miny > it->pixelMinY) 	miny = it->pixelMinY;
		if (maxy < it->pixelMaxY)	maxy = it->pixelMaxY;
	}


	// use the world coordinate limits <maxx, minx, maxx, minx> to determine the
	// sized  of the bitmap buffer to allocate for drawing the image
	// always make image imageWidth multiple of 8
    unsigned imageWidth 	= unsigned(ceil ( (maxx - minx) + 2*optBoarder + 1 ));
    unsigned imageHeight	= unsigned(ceil ( (maxy - miny) + 2*optBoarder + 1 ));
    int xOffset		= int(floor( optBoarder ));
    int yOffset		= xOffset;

    bool isPolarityDark = true;
    isPolarityDark = (optInvertPolarity ^ gerbers.front()->imagePolarityDark);	// polarity is relative to 1st gerber file
    if ( rowsPerStrip > unsigned(imageHeight) || rowsPerStrip == 0)
    	rowsPerStrip = imageHeight;
	unsigned darkPixelsCount = 0;


    //
    // Eye candy
    //
    if (optVerbose >= 2)
    {
    	printf("polygon count:               %d\n",globalPolygons.size());
    	printf ("grow option:                 %.1f pixels , %.3f mm\n", optGrowSize, optGrowSize/imageDPI*25.4);
    }
    if (optVerbose >= 1)
    {
		printf ("Image data\n"
				"  origin (mm):               %.3f x %.3f\n"
				"  size (mm):                 %.3f x %.3f\n"
				"  size (pixels):             %u x %d\n"
				"  uncompressed size (MB):    %.1f\n"
				"  dots per inch:             %u\n"
				"  TIFF rows per strip        %u\n"
				,(-xOffset+minx)/imageDPI*25.4, (-yOffset+miny)/imageDPI*25.4
				,imageWidth/imageDPI*25.4, imageHeight/imageDPI*25.4
				,imageWidth, imageHeight
				,float( (((imageWidth+7) / 8) * imageHeight) / 0x100000)
				,int(imageDPI)
				,rowsPerStrip);

	}
    fflush(stdout);

	if (optTestOnly)	// stop here on testing or if no Polygons to draw
	{
		if (optVerbose)
			printf("  time (sec):                %.2f\n",((double) (clock() - start_clock)) / CLOCKS_PER_SEC );
		return 0;
	}

	// Initialise TIFF with the libtiff library
	//
	TIFF* tif = TIFFOpen(outputFilename.c_str(), "w");
    if	(tif==NULL)
    {
    	cout << "error creating output file '" << outputFilename << "\n";;
    	return 1;
    }

    TIFFSetField(tif, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG);		// avoid errors, dispite TIFF spec saying this tag not needed in monochrome images.
    TIFFSetField(tif, TIFFTAG_PHOTOMETRIC, PHOTOMETRIC_MINISWHITE);		// white pixels are zero
    TIFFSetField(tif, TIFFTAG_COMPRESSION, COMPRESSION_CCITTRLE);		// use CCITT Group 3 1-Dimensional Modified Huffman run length encoding
    TIFFSetField(tif, TIFFTAG_IMAGELENGTH, imageHeight);
    TIFFSetField(tif, TIFFTAG_IMAGEWIDTH, imageWidth);
    TIFFSetField(tif, TIFFTAG_RESOLUTIONUNIT, 2);					// Resulution unit in inches
    TIFFSetField(tif, TIFFTAG_YRESOLUTION, imageDPI);
    TIFFSetField(tif, TIFFTAG_XRESOLUTION, imageDPI);
    TIFFSetField(tif, TIFFTAG_ROWSPERSTRIP, rowsPerStrip);

	//
    // Calculate size and allocate buffer for drawing. The image will be rendered sequential blocks of
    // imageWidth wide by rowsPerStrip high.
    //
	bytesPerScanline = ((imageWidth+7) >> 3);
	bitmapBytes = bytesPerScanline * rowsPerStrip;
    bitmap = (unsigned char *)malloc( bitmapBytes);
    if ( bitmap == 0 )
    	error("cannot allocate memory");


    //-----------------------------------------------------------------------
    // Draw polygons
    //-----------------------------------------------------------------------
    xOffset -= minx;


    int stripCounter = 0;
	list<Polygon>::iterator polyIterator = globalPolygons.begin();
    list<PolygonReference >  activePolys;

	// The bitmap will be divided into strips, of height rowsPerStrip.
	// Polygons are plotted for each strip consecutively in a loop, where the strip y coordinate equals ystart
    for (int ystart = miny - yOffset; ystart < ( int(imageHeight) + miny - yOffset); ystart += rowsPerStrip)
    {
    	// blank entire strip buffer, set pixels on/off depending on polarity of the 1st Gerber.
	    if (isPolarityDark)	memset(bitmap, 0x00, bitmapBytes);
	    else				memset(bitmap, 0xff, bitmapBytes);

	    unsigned char *bufferLine = bitmap;

	    // Loop over each row of the strip and fill with horizontal lines from the polygon raster data.
	    // All polygon are sorted in the list globalPolygons. Iterating each polygon for raster data will guarantee no missing lines.
		for (int y = ystart; (y-ystart) < rowsPerStrip && (y <= maxy); y++ , bufferLine += bytesPerScanline)
		{
			while (polyIterator != globalPolygons.end() && y == (polyIterator->pixelMinY))
			{
				activePolys.push_back( PolygonReference() );
				activePolys.back().polygon = &(*polyIterator);
				activePolys.sort();
//				printf("added poly %d (y=%d)\n", activePolys.back().polygon->number, y);
				polyIterator++;
			}

			for (list<PolygonReference>::iterator it = activePolys.begin();  it != activePolys.end();)
			{
				if (y > it->polygon->pixelMaxY)
				{
//					printf("erased poly %d (y=%d)\n", activePolys.back().polygon->number, y);
					it = activePolys.erase(it);
					continue;
				}
				int sliCount;
				int *sliTable;
				it->polygon->getNextLineX1X2Pairs( sliTable, sliCount);

//				printf("p %2d y:%d (x cnt %d) |",it->polygon->number, y, sliCount); fflush(stdout);

				Polarity_t pol =  it->polygon->polarity;
				if ((pol == DARK) && !isPolarityDark) pol = CLEAR;
				if ((pol == CLEAR) && isPolarityDark) pol = DARK;

				for (int i=0; i < sliCount; i+=2)
				{
//					printf(" %d~%d ",sliTable[i], sliTable[i+1] ); fflush(stdout);
					horizontalLine( xOffset + it->polygon->pixelOffsetX + sliTable[i], \
									xOffset + it->polygon->pixelOffsetX + sliTable[i+1], \
									bufferLine, pol  );

				}
//				printf("\n");
				it++;
			}
        }

		//
		// Write strip buffer to TIFF
		//
        unsigned lines = min (rowsPerStrip, imageHeight - rowsPerStrip * stripCounter);
        int percentComplete = (100*(stripCounter*rowsPerStrip + lines))/imageHeight;
	    if (optVerbose)
		{
	    	static int last = percentComplete;
	    	if (percentComplete != last )
	    		cout << "Rendering "<< percentComplete <<"%  \r"<<flush;
	    	last = percentComplete;
		}
    	TIFFWriteEncodedStrip(tif, stripCounter++, bitmap, bytesPerScanline*lines);

    	// Calculate positive area information
        if (optShowArea)
        {
			for (int i=0; i < lines; i++)
			{
				unsigned char *pbitmaprow = bitmap + bytesPerScanline * i;
				for (int x=0; x < bytesPerScanline; x++)
					darkPixelsCount += nbitsTable [ *pbitmaprow ];
					pbitmaprow++;
			}
        }

    }
    TIFFClose(tif);

    if (optVerbose)    	cout << "\n";

    if (optShowArea)
    {
    	printf("  dark  area (sq.cm):        %0.1f\n",darkPixelsCount*2.54*2.54/(imageDPI*imageDPI));
    	printf("  clear area (sq.cm):        %0.1f\n",((imageHeight*imageWidth) - darkPixelsCount*2.54*2.54)/(imageDPI*imageDPI));
    }

	if (optVerbose)
		printf("  time (sec):                %.2f\n",((double) (clock() - start_clock)) / CLOCKS_PER_SEC );

}



